packages/runtime/container-runtime/src/summary/runningSummarizer.ts

```diff
@@ -664,6 +661,139 @@ export class RunningSummarizer implements IDisposable {
 		});
 	}
 
+	/**
+	 * Tries to summarize with dynamic retries, i.e., retry is based on the failure params and error received.
+	 * For instance, if summarize failed due to a server error that has `retryAfterSeconds` set, wait for that duration
+	 * and retry summarize. Similarly, if summary generation failed but the error is retriable, retry before bailing out.
+	 */
+	private async trySummarizeWithDynamicRetries(
+		reason: SummarizeReason,
+		cancellationToken = this.cancellationToken,
+	) {
+		// Currently, we try 2 overall attempts with retries in each attempt. The second attempt will close the
+		// summarizer at the summarization call site but we need to keep this code for the time being. This will be
+		// reduced to a single attempt.
+		const attemptOptions: ISummarizeOptions[] = [
+			{ refreshLatestAck: false, fullTree: false },
+			{ refreshLatestAck: true, fullTree: false },
+		];
+		for (const summarizeOptions of attemptOptions) {
+			let previousStage: SummaryStage | undefined;
+
+			// The maximum number of retries across failure stages. If summarization does not succeed by these number
+			// of retries, quit and close the summarizer.
+			const maxRetries = defaultMaxRetries;
+			let currentRetries = -1;
+
+			// The maximum number of retries for a given stage. The default value is 1 but this can be set via failure
+			// params. Each failure is different and some can be succeeded by retrying more often than others.
+			// If summarization fails at different stages during retries, each stage gets some retries. However, the
+			// total number of retries across stages cannot exceed `maxRetries`.
+			let maxRetriesForStage = 1;
+			let currentRetriesForStage = -1;
+
+			type ShouldRetryResult =
+				| { retry: false }
+				| {
+						retry: true;
+						stage: SummaryStage;
+						delayMs: number;
+				  };
+			// Returns whether summarization should be retried based on the failure data.
+			const shouldRetry = (
+				failureData: ISubmitSummaryFailureResult | undefined,
+			): ShouldRetryResult => {
+				// For lastSummary, do not retry.
+				// If the failure data does not specify `retryAfterSeconds` or the stage at which summarization failed,
+				// do not retry.
+				// If maxRetries have been exceeded, do not retry anymore.
+				if (
+					reason === "lastSummary" ||
+					failureData === undefined ||
+					failureData.retryAfterSeconds === undefined ||
+					failureData.stage === undefined ||
+					currentRetries >= maxRetries
+				) {
+					return { retry: false };
```

Comment:
Should we have a stage here as base?

Reply:
Why?

Reply:
Wouldn't it be useful to know what stage we failed at if we couldn't retry?

Reply:
Good catch. This stage is logged later and it will be missing when retry is false.

Reply:
Updated.